File: pjecalc-fonte/negocio/br/jus/trt8/pjecalc/negocio/dominio/calculo/atualizacao/regras/DataEspecificaJurosMoraParametrosAtualizacaoValidRule.java
/*
 * Decompiled with CFR 0.152.
 */
package br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.regras;

import br.jus.trt8.pjecalc.base.comum.HelperDate;
import br.jus.trt8.pjecalc.base.comum.Utils;
import br.jus.trt8.pjecalc.base.comum.validadores.ValidatorContext;
import br.jus.trt8.pjecalc.negocio.comum.validators.ValidRule;
import br.jus.trt8.pjecalc.negocio.comum.validators.ValidValueValidator;
import br.jus.trt8.pjecalc.negocio.constantes.Mensagens;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.ExcecaoDeJurosDaAtualizacao;
import br.jus.trt8.pjecalc.negocio.dominio.calculo.atualizacao.ParametrosDeAtualizacao;
import java.util.Date;

public class DataEspecificaJurosMoraParametrosAtualizacaoValidRule
implements ValidRule {
    private static final String DATA_HOJE = "dataHoje";
    private static final String DATA_AJUIZAMENTO = "calculo.dataAjuizamento";
    private int flag = 0;

    @Override
    public Mensagens getMessage() {
        if (this.flag == 0) {
            return Mensagens.MSG0007;
        }
        if (this.flag == 1) {
            return Mensagens.MSG0010;
        }
        return Mensagens.MSG0008;
    }

    @Override
    public boolean isValid(ValidValueValidator validator, ValidatorContext context) {
        ParametrosDeAtualizacao parametrosDeAtualizacao;
        boolean ehDaExcecaoJuros = false;
        if (context.getBean() instanceof ParametrosDeAtualizacao) {
            parametrosDeAtualizacao = (ParametrosDeAtualizacao)context.getBean();
        } else {
            parametrosDeAtualizacao = ((ExcecaoDeJurosDaAtualizacao)context.getBean()).getParametrosDeAtualizacao();
            ehDaExcecaoJuros = true;
        }
        if (Utils.nulo(parametrosDeAtualizacao.getCalculo())) {
            return true;
        }
        Date apartirDe = (Date)context.getValue();
        Date ajuizamento = parametrosDeAtualizacao.getCalculo().getDataAjuizamento();
        if (Utils.nulo(apartirDe) || Utils.nulo(ajuizamento)) {
            return true;
        }
        if (ehDaExcecaoJuros) {
            if (HelperDate.dateBefore(apartirDe, ajuizamento)) {
                this.flag = 2;
                validator.getParameters().put("1", DATA_AJUIZAMENTO);
                return false;
            }
        } else if (HelperDate.dateBeforeOrEquals(apartirDe, ajuizamento)) {
            validator.getParameters().put("1", DATA_AJUIZAMENTO);
            return false;
        }
        if (!ehDaExcecaoJuros) {
            HelperDate hoje = HelperDate.getInstance();
            hoje.removeTime();
            if (hoje.lessThen(apartirDe)) {
                this.flag = 1;
                validator.getParameters().put("1", DATA_HOJE);
                return false;
            }
        }
        return true;
    }
}

File: src/lib/pjecalc/core/dominio/calculo/atualizacao/regras/data-especifica-juros-mora-parametros-atualizacao-valid-rule.ts
/**
 * Porte 1:1 de DataEspecificaJurosMoraParametrosAtualizacaoValidRule.java (73 linhas).
 *
 * Valida a data "aPartirDe" para juros de mora:
 *   - ParametrosDeAtualizacao: data deve ser > ajuizamento e <= hoje.
 *   - ExcecaoDeJurosDaAtualizacao: data deve ser >= ajuizamento.
 *
 * Ref: pjecalc-fonte/.../dominio/calculo/atualizacao/regras/DataEspecificaJurosMoraParametrosAtualizacaoValidRule.java
 */
import { HelperDate } from '../../../../base/comum/helper-date';
import { ExcecaoDeJurosDaAtualizacao } from '../excecao-de-juros-da-atualizacao';
import { ParametrosDeAtualizacao } from '../parametros-de-atualizacao';

export interface JurosMoraValidResult {
  valid: boolean;
  message?: string;
  detailParam?: string;
}

export class DataEspecificaJurosMoraParametrosAtualizacaoValidRule {
  static isValid(
    aPartirDe: Date | null,
    bean: ParametrosDeAtualizacao | ExcecaoDeJurosDaAtualizacao,
  ): JurosMoraValidResult {
    let parametros: ParametrosDeAtualizacao | null;
    let ehDaExcecaoJuros = false;
    if (bean instanceof ParametrosDeAtualizacao) {
      parametros = bean;
    } else {
      parametros = bean.getParametrosDeAtualizacao();
      ehDaExcecaoJuros = true;
    }

    if (!parametros || !parametros.getCalculo()) return { valid: true };
    const calc = parametros.getCalculo() as unknown as { getDataAjuizamento?: () => Date | null };
    const ajuizamento = calc.getDataAjuizamento?.();

    if (!aPartirDe || !ajuizamento) return { valid: true };

    if (ehDaExcecaoJuros) {
      if (HelperDate.dateBefore(aPartirDe, ajuizamento)) {
        return { valid: false, message: 'MSG0008', detailParam: 'calculo.dataAjuizamento' };
      }
    } else if (HelperDate.dateBeforeOrEquals(aPartirDe, ajuizamento)) {
      return { valid: false, message: 'MSG0007', detailParam: 'calculo.dataAjuizamento' };
    }

    if (!ehDaExcecaoJuros) {
      const hoje = new Date();
      hoje.setHours(0, 0, 0, 0);
      if (HelperDate.dateBefore(hoje, aPartirDe)) {
        return { valid: false, message: 'MSG0010', detailParam: 'dataHoje' };
      }
    }
    return { valid: true };
  }
}
